src/ledger/LedgerTxnImpl.h

```diff
@@ -421,5 +421,17 @@ class LedgerTxnRoot::Impl
     // - the prepared statement cache may be, but is not guaranteed to be,
     //   modified
     void writeSignersTableIntoAccountsTable();
+
+    // encodedDataNamesBase64 has the basic exception safety guarantee. If it
+    // throws an exception, then
+    // - the prepared statement cache may be, but is not guaranteed to be,
+    //   modified
+    void encodeDataNamesBase64();
```

Comment:
Do we need those in that class? Woudn't static methods/free function be better here? Those are one-time calls (even one-time per instance, not per launch), so why pollute that interface?

Reply:
OK, but I don't agree with that solution as I would like to see a layer below LedgerTxnRoot to handle the queries.

For example dropOffers does a lot of unnecessary things, like "throwIfChild" (it is only called from Database::initialize!). I think we could make it as safe as it is now with a different solution (but this is very low priority).

And if you really want all SQL statements in LedgerTxnRoot, why not move there also all of `ALTER TABLE` from liabilities and signers changes?

Reply:
yeah I would like to see some separation too (and/or limit public methods) but that is outside of the scope of this PR and seems like lower priority. Open an issue to discuss that if you feel like we should push for it.